app/client/components/forms/BasicForm.tsx

```diff
@@ -0,0 +1,38 @@
+import React from "react";
+
+import { CSRFToken, FieldHandler, FormHandler } from "@reactivated";
+import { FieldMap } from "reactivated/dist/forms";
+import { Alert } from "../Alert";
+import { BasicFormField } from "./BasicFormField";
+
+interface BasicFormProps<T extends FieldMap> {
+  form: FormHandler<T>;
+  submitButtonLabel: string;
+}
+
+export const BasicForm = <T extends FieldMap>({
+  form,
+  submitButtonLabel,
+}: BasicFormProps<T>) => {
+  return (
+    <form method="POST">
+      <CSRFToken />
+      {form.nonFieldErrors?.map((error) => (
+        <Alert key={error} type="error">
+          {error}
+        </Alert>
+      ))}
+      {form.visibleFields.map((field) => (
+        <BasicFormField
+          key={field.name}
+          // @ts-expect-error Typescript complains about this conversion, but it can't be avoided without changing types exposed by Reactivated.
```

Comment:
We are losing some information by making this component generic. That's why TypeScript is angry: we cannot be sure that this matching will work. I am unfamiliar with the library, so I don't know if this matching won't work. If you are sure, then that's fine, but I'd resign from making this generic. I suggest removing this component and doing all the logic in the login form. Sometimes, being too generic might bite us, and when we are forced to ignore TypeScript, that might be the case.

But it's up to you if you are sure about this casting.

Reply:
I strongly prefer to have a generic `BasicForm` component to not have to write all the boilerplate form code in each view. This way whenever you define a from in Django, all of the fields will be automatically rendered on the frontend side (which means that if you for example add a field to a Django form, no frontend changes will be required). In fact that's how it currently works with this library: https://github.com/ksiuwr/django-materializecss-form

That said, for some forms more flexibility will be required and then it makes sense to manually specifiy all the fields on the frontend side. This will be the case with register form, which I will publish in the next PR, so it might make more sense then ;)

I'm sure about this casting. Under the hood `FieldHandler` is just a more generic type than the type of `field`. The reason why it's a problem is because Reactivated expects each form field to be specified manually, rather than iterating through `form.visibleFields`. [Here is example from docs](https://www.reactivated.io/documentation/api/#useform). 

But again, making this generic will make life way easier, considering that zosia website has a lot of forms defined in Django.